### PR TITLE
DQL custom functions: document TypedExpression

### PR DESCRIPTION
**Pull request by janedbal**
Partially related to https://github.com/doctrine/orm/issues/11537

### CODE REVIEW

File: docs/en/cookbook/dql-user-defined-functions.rst [hunk 1]
```diff
@@ -232,6 +232,33 @@ vendors SQL parser to show us further errors in the parsing
 process, for example if the Unit would not be one of the supported
 values by MySql.
 
+Typed functions
+----------
```
Comment (SenseException) [inline docs/en/cookbook/dql-user-defined-functions.rst]:
```suggestion
Typed functions
---------------
```

File: docs/en/cookbook/dql-user-defined-functions.rst [hunk 2]
```diff
@@ -232,6 +232,33 @@ vendors SQL parser to show us further errors in the parsing
...
+By default, result of custom functions is fetched as-is from the database driver.
+If you want to be sure the type is always the same, your custom function can
```
Comment (SenseException) [inline docs/en/cookbook/dql-user-defined-functions.rst]:
```suggestion
If you want to be sure that the type is always the same, then your custom function needs to 
```

File: docs/en/cookbook/dql-user-defined-functions.rst [hunk 3]
```diff
@@ -232,6 +232,33 @@ vendors SQL parser to show us further errors in the parsing
 process, for example if the Unit would not be one of the supported
 values by MySql.
 
+Typed functions
+----------
+By default, result of custom functions is fetched as-is from the database driver.
+If you want to be sure the type is always the same, your custom function can 
+implement ``Doctrine\ORM\Query\AST\TypedExpression``. Then, the result is wired
+through ``convertToPhpValue`` method of the ``Type`` returned from ``getReturnType``.
```
Comment (SenseException) [inline docs/en/cookbook/dql-user-defined-functions.rst]:
```suggestion
through the ``Doctrine\DBAL\Types\Type::convertToPhpValue()`` method returned in ``getReturnType``.
```

**Review by SenseException — CHANGES_REQUESTED** [review]

File: docs/en/cookbook/dql-user-defined-functions.rst (hunk 3 above)
Comment (janedbal) [reply docs/en/cookbook/dql-user-defined-functions.rst]: This suggestion feels more weird to me. No method is returned (obviously), Type is returned.

Comment (SenseException) [reply docs/en/cookbook/dql-user-defined-functions.rst]: I didn't expect that someone could read that a "method" is returned.

```suggestion
through ``Doctrine\DBAL\Types\Type::convertToPhpValue()`` of the ``Type`` returned in ``getReturnType()``.
```

**Review by SenseException — COMMENTED** [review]
:+1: Please [squash](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing) your 4 commits into 1.

**Comment by janedbal** [comment]
No problem. Just out of curiosity, why do you force contributors to squash it manually when you can do it with a single button?

image

**Comment by greg0ire** [comment]
For me, there are at least 2 reasons:
- Pressing that button forces us to come up with a commit message summarizing all the commits. Crafting that commit message should IMO be your job, since it's your changes.
- The commit hash stays the same before and after merging, when no squashing is involved. That can be handy when trying to figure out what branch contains a particular commit or whatnot.

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @janedbal !

**Pull request merged**